### PR TITLE
Lint compiled HTML

### PR DESCRIPTION
**Pull request by domoscargin**
We currently have no built-in way of ensuring our compiled HTML changes in the way we expect, and doesn't break.

This PR adds HTML linting using `html-validate`. There are quite a few commits, but they're all really small, promise!

There are some issues to consider:

- I've ignored a coupla "WCAG techniques" type rules which we might want to review as part of our accessibility work ([495be15](https://github.com/alphagov/govuk-design-system/pull/2569/commits/4bc4910b110dfe170bbce3d66f02037d47a451c1) and [d1dea99](https://github.com/alphagov/govuk-design-system/pull/2569/commits/d1dea99330909005c43473a4521434f8e70ef0b6))
- I've tried to avoid inline validation as far as possible, but sometimes this was the most targetted way
- We currently have status labels contained within `<h1>` tags, which introduces invalid child elements. See #2606 
- I've popped in a temporary inline fix for #2260 until that's fixed
- I've allowed nested forms for the few cases where that's an issue: #2609. Chrome, at least, simply rips out the inner `<form>` elements, but it's still messy

### CODE REVIEW

**Comment by querkmachine** [comment]
Do you have a sample of what problems it's finding? Nearly 2000 issues seems like an excessive number!

Also I imagine there are examples of invalid code where things are wrong-on-purpose because they have better outcomes for the end user (e.g. [our use of aria-attributes on conditional checkboxes/radios](https://github.com/alphagov/govuk-frontend/pull/2823)). We probably don't want to treat those exceptionally.

**Comment by domoscargin** [comment]
I haven't delved too deeply, but the list of rules broken:
```
https://html-validate.org/rules/void-style.html
https://html-validate.org/rules/wcag/h30.html
https://html-validate.org/rules/aria-label-misuse.html
https://html-validate.org/rules/attr-quotes.html
https://html-validate.org/rules/no-implicit-close.html
https://html-validate.org/rules/element-permitted-content.html
https://html-validate.org/rules/attr-case.html
https://html-validate.org/rules/no-deprecated-attr.html
https://html-validate.org/rules/no-dup-id.html
https://html-validate.org/rules/attribute-allowed-values.html
https://html-validate.org/rules/wcag/h32.html
https://html-validate.org/rules/long-title.html
https://html-validate.org/rules/attr-delimiter.html
https://html-validate.org/rules/wcag/h63.html
https://html-validate.org/rules/form-dup-name.html
https://html-validate.org/rules/close-order.html
```
There's a lot of what I'm pretty sure is noise:

`aria-label-misuse` because we've got aria-labels on our `app-navigation__subnav` `ul` elements, which I assume is a deliberate choice.
`void-style` errors mostly seem annoyed about us self-closing `<link>` elements which feels heavy-handed.
`wcag/h30` on our top anchor: `<a id="top"></a>`

Probably legit errors include:

Most of the `attr-quotes` look like they could be switched to double quotes with no issue.
`no-implicit-close`: there's a few unclosed `<p>` tags floating about: `<p>Share your experience, research and skills to help us reiterate the Task list pages pattern into a component.<p>`

🤷 
`element-permitted-content`: Doesn't like the "experimental" `div` contained in the `h1` on the Resources and Tools page

Etc, etc.

So I'm assuming 90% of the errors could safely be ignored, but haven't really had a close look.

**Comment by colinrotherham** [comment]
@domoscargin I think this it great 🙌 

Gives us more confidence that Markdown/Nunjucks is outputting what we want

**Comment by domoscargin** [comment]
I've created an issue to deal with some invalid HTML here: https://github.com/alphagov/govuk-design-system/issues/2606

Down to 65 errors to check.

File: package.json
Comment (colinrotherham) [inline package.json:23]: Shall we make this `lint:html` and get it running on GitHub Actions? 🙌 

```suggestion
    "lint:html": "html-validate \"deploy/public/**/index.html\""
```

Or alternatively could set up an npm "post" script?

```suggestion
    "postbuild": "html-validate \"deploy/public/**/index.html\""
```

**Comment by domoscargin** [comment]
I've added another issue to deal with nested form elements here: #2609 

In addition, #2260 is causing `form-dup-name` errors. This wouldn't happen if the rule checked for unique values, as suggested by @colinrotherham: https://github.com/alphagov/govuk-frontend/pull/3270#pullrequestreview-1290744193, but I feel like at a certain point, the rule becomes a bit toothless if we're basically ignoring all its results...

I could finagle this PR to ignore these errors (like it ignores #2606), but tempted to instead block this on these other issues.

Once #2606, #2609 and #2260 are fixed, there'll be no more HTML validation errors and we can tweak this PR and get it merged.

File: package.json
Comment (domoscargin) [reply package.json:23]: Hmm, neither seems fully satisfying to me - having it in Github Actions means we don't catch things locally, but setting up a `postbuild` script means it won't run on `npm run test`.

And adding it to `test` means we'll have to include the 8-10 second build step for that script, which is also annoying.

Comment (colinrotherham) [reply package.json:23]: Ahh, but it would still be lovely to see all your hard work in GitHub Actions

Could always call into the task from multiple places where needed?

```json
"pretest": "npm run validate"
"postbuild": "npm run validate"
```

Possibly a daft example 🙈

Comment (domoscargin) [reply package.json:24]: For now, I've just made the `lint:html` script. I briefly had it running in Github actions, but it was having parsing errors which I'll try and figure out.

File: .htmlvalidateignore
Comment (colinrotherham) [inline .htmlvalidateignore:1]: Mind sticking a new line at the end here?

Usually an EditorConfig job!

File: package.json
Comment (colinrotherham) [reply package.json:24]: @domoscargin Would you mind if we resolved the GitHub Actions checks before merging?

Would be great to validate all my Markdown and code block changes automatically

File: views/partials/_navigation.njk
Comment (colinrotherham) [inline views/partials/_navigation.njk:14]: Have we got an issue for this one?

File: package.json
Comment (domoscargin) [reply package.json:24]: Yeah, sorry, I should've popped this back to draft maybe - something breaks the Github Actions and I haven't had time to dig into what it is. Some kind of parsing maybe.

File: views/partials/_navigation.njk
Comment (domoscargin) [reply views/partials/_navigation.njk:14]: It was a deliberate choice for accessibility: https://github.com/alphagov/govuk-design-system/commit/4ff06e4961bcd4a147d580d1d5ac2736fe9f6ec3

@36degrees dya reckon I should make an issue to review this?

File: package.json
Comment (domoscargin) [reply package.json:24]: https://github.com/alphagov/govuk-design-system/actions/runs/4222561918/jobs/7331207893#step:7:28

Comment (domoscargin) [reply package.json:24]: (Got some time set aside to have a look at this later, but if it's an obvious fix...)

File: .htmlvalidateignore
Comment (domoscargin) [reply .htmlvalidateignore]: nice catch, thanks!

VS Code seems to be ignoring my .editorconfig file. I'll have to dig around and figure out why.

File: package.json
Comment (colinrotherham) [reply package.json:24]: Ah it was an invalid config at our end which hit a bug in `html-validate`

They're using AJV for JSON schema validation, but _maaybe_ it's not working with our JS config?

If you log the config validation error before it's formatted you get:

>SchemaValidationError [UserError]: Rule configuration error: /rules/form-dup-name/1/shared/3 must be equal to one of the allowed values: radio, checkbox, submit

Comment (domoscargin) [reply package.json:24]: Could there be an outdated version of html-validate cached?

Comment (domoscargin) [reply package.json:24]: Ah, no! My bad - I haven't updated my package.json.

Comment (domoscargin) [reply package.json:24]: Yup, just needed to bump `html-validate` to get the updated rule and it's all working now. Thanks for the helpful error log wrangling!

File: .github/workflows/test.yaml
Comment (colinrotherham) [inline .github/workflows/test.yaml:52]: Not a blocker, but confused me that `npm run lint` doesn't include `lint:html`

Should we copy the GOV.UK Frontend approach?

1. Run all lint tasks with `npm run lint`
2. Tests trigger build via `pretest`
3. Checks that need a build use `postbuild`

**Review by colinrotherham — APPROVED** [review]

File: .github/workflows/test.yaml
Comment (domoscargin) [reply .github/workflows/test.yaml:52]: It's not as scalable, but for now I've just added a `prelint:html` script that builds. That should run on `test`, `lint` and `lint:html` locally and during CI.

We can revisit if we ever decide we need to do anything more fancy.

Comment (domoscargin) [reply .github/workflows/test.yaml]: Forgive my cack-handedness at Github Actions!

I've removed the separate build step, but made it clear we're still building as part of the test step.

File: views/partials/_navigation.njk
Comment (36degrees) [reply views/partials/_navigation.njk:14]: I'm not sure where the list of 'allowed elements' on https://html-validate.org/rules/aria-label-misuse.html comes from… according to https://www.w3.org/TR/html-aria/#el-li `li` is not listed as ['Naming Prohibited'](https://www.w3.org/TR/html-aria/#dfn-naming-prohibited) and the [`listitem` role](https://www.w3.org/TR/wai-aria-1.2/#listitem) inherits the `aria-label` property… 🤔 

The rationale in the linked commit makes sense to me, but I think given it's been flagged it might be worth asking @davidc-gds to review it just in case? Whether or not we need an issue for that I'm unsure – maybe ask David?

Comment (domoscargin) [reply views/partials/_navigation.njk:14]: Interesting. Might dig about a bit in the `html-validate` source code and figure out if it's worth filing a bug. Will speak to David as well.

Comment (domoscargin) [reply views/partials/_navigation.njk:14]: Ah, so the `<ul>` text reads:

> [Global aria-* attributes](https://www.w3.org/TR/wai-aria-1.2/#global_states) and any aria-* attributes **applicable to the allowed roles**.

Which is sort of ambiguous, but I think basically means we need to apply a role to the `<ul>` element - if we do that, then `html-validate` is happy.

By default, the `<ul>` has the `list` role, which is:

> allowed, but NOT RECOMMENDED.

I guess the `menu` role feels most appropriate here?

A bit confusingly, in the case where this applies (when the mobile menu is visible), we automatically expand the "Community" sub-menu, but list the "Community" button as collapsed. It's arguably fine since it's just a set of links, but also maybe a bit weird.

Comment (domoscargin) [reply views/partials/_navigation.njk:14]: Basically, I think this could do with an issue 😂 

I'll make one

Comment (36degrees) [reply views/partials/_navigation.njk:14]: Aah, of course – not sure why I was looking at `<li>`!

> By default, the `<ul>` has the `list` role, which is:
> 
> > allowed, but NOT RECOMMENDED.

I _believe_ when they say it's 'allowed but not recommended' they mean it's not recommended to set this role _explicitly_ – if you take a look at the headings for the table it says "Implicit ARIA semantics (explicitly assigning these in markup is NOT RECOMMENDED)".

> I guess the `menu` role or `navigation` feels most appropriate here? Though it's already contained within a `nav`.

We should avoid the navigation role for exactly that reason – it's already in a nav region. The [`menu` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role) is designed for interfaces that mimic desktop app style interfaces where you have menus with lists of items, like in a 'menu bar' and wouldn't be appropriate here.

`list` still feels like the appropriate role – I think if we changed it, we'd lose the information like 'List, 10 items' that screen reader users get which helps them understand how many links there are, for example.

Still worth having an issue to explore further, but I think either 1) having an `aria-label` is fine on a `<ul>` and is useful for users or 2) we should just remove the `aria-label`. I don't think changing the role makes sense.

Comment (domoscargin) [reply views/partials/_navigation.njk:14]: I've added an issue and linked to this discussion.

https://github.com/alphagov/govuk-design-system/issues/2624

**Pull request merged**
